File: bots/rs/v55/src/util/visualiser.rs
//! Translation of `bots/intgrah/v54.7.9/util/visualiser.py`.
//!
//! Visualisation primitives. Defines the `Dump` enum (each variant a typed
//! payload the viewer knows how to render) and the `Dumper` struct that bots
//! use to emit a per-turn JSON tree of named dump nodes.

use cambc::Position;
use serde::Serialize;
use std::collections::HashMap;

#[derive(Clone, Copy, Debug, Serialize)]
pub struct Colour {
    pub r: u8,
    pub g: u8,
    pub b: u8,
    pub a: u8,
}

impl Colour {
    #[must_use]
    pub const fn new(r: u8, g: u8, b: u8, a: u8) -> Self {
        Self { r, g, b, a }
    }

    fn as_array(&self) -> [u8; 4] {
        [self.r, self.g, self.b, self.a]
    }
}

pub const TRANSPARENT: Colour = Colour::new(0, 0, 0, 0);

/// One stop in a `Palette`, generic over the scalar type `T` (bool / int / float).
#[derive(Clone, Debug, Serialize)]
pub struct PaletteStop<T> {
    pub t: T,
    pub colour: Colour,
}

/// Linearly interpolated colour palette over scalar values of type `T`.
#[derive(Clone, Debug, Serialize)]
pub struct Palette<T> {
    pub stops: Vec<PaletteStop<T>>,
    pub special: Vec<(T, Colour)>,
}

impl<T> Palette<T> {
    #[must_use]
    pub fn new(stops: Vec<PaletteStop<T>>) -> Self {
        Self {
            stops,
            special: Vec::new(),
        }
    }
}

impl<T: Clone + PartialEq> Palette<T> {
    /// Return a copy of the palette with additional special values merged in.
    #[must_use]
    pub fn with_special(&self, special: &[(T, Colour)]) -> Self {
        let mut merged: Vec<(T, Colour)> = self.special.clone();
        for (k, c) in special {
            if let Some(slot) = merged.iter_mut().find(|(kk, _)| kk == k) {
                slot.1 = *c;
            } else {
                merged.push((k.clone(), *c));
            }
        }
        Self {
            stops: self.stops.clone(),
            special: merged,
        }
    }
}

// Pre-built palettes. Python defines these as module constants; Rust needs
// `Vec`s so they are functions instead.
#[must_use]
pub fn green_red() -> Palette<i64> {
    Palette::new(vec![
        PaletteStop {
            t: 0,
            colour: Colour::new(50, 200, 50, 140),
        },
        PaletteStop {
            t: 100,
            colour: Colour::new(200, 50, 50, 140),
        },
    ])
}

#[must_use]
pub fn blue_red() -> Palette<i64> {
    Palette::new(vec![
        PaletteStop {
            t: 0,
            colour: Colour::new(50, 50, 200, 140),
        },
        PaletteStop {
            t: 100,
            colour: Colour::new(200, 50, 50, 140),
        },
    ])
}

#[must_use]
pub fn fog() -> Palette<bool> {
    Palette::new(vec![
        PaletteStop {
            t: false,
            colour: TRANSPARENT,
        },
        PaletteStop {
            t: true,
            colour: Colour::new(0, 0, 0, 180),
        },
    ])
}

// ---------------------------------------------------------------------
// Dump variants: every value passed to `Dumper::dump` is one of these.
// ---------------------------------------------------------------------

/// A scalar value carried by a `DumpScalar` node.
#[derive(Clone, Debug, Serialize)]
#[serde(untagged)]
pub enum ScalarValue {
    Int(i64),
    Float(f64),
    Bool(bool),
    Str(String),
    Null,
}

/// All dumpable payload types. Each variant matches one Python `Dump*`
/// dataclass; serialisation produces the exact `{"$type": "...", ...}` shape
/// the Python viewer expects.
#[derive(Clone, Debug)]
pub enum Dump {
    BoolGrid {
        data: Vec<bool>,
        palette: Palette<bool>,
    },
    U8Grid {
        data: Vec<u8>,
        palette: Palette<i64>,
    },
    I16Grid {
        data: Vec<i16>,
        palette: Palette<i64>,
    },
    U16Grid {
        data: Vec<u16>,
        palette: Palette<i64>,
    },
    F32Grid {
        data: Vec<f32>,
        palette: Palette<f64>,
    },
    /// Unordered set of tiles, rendered as cell rings.
    Tiles { data: Vec<Position> },
    /// Single tile rendered as a ring around the cell. `pos = None` renders
    /// as nothing so callers can dump unconditionally.
    Tile { pos: Option<Position> },
    /// Single tile rendered as a filled coloured dot.
    Dot {
        pos: Option<Position>,
        colour: Colour,
    },
    /// Ordered list of tiles rendered as a polyline.
    Path {
        points: Vec<Position>,
        colour: Colour,
    },
    VectorField {
        angles: Vec<Option<f64>>,
        magnitudes: Option<Vec<f64>>,
    },
    /// Plain value (int / float / bool / str / None) shown verbatim.
    Scalar { value: ScalarValue },
}

fn serialise_palette_t<T: Serialize + Clone>(p: &Palette<T>) -> serde_json::Value {
    let stops: Vec<serde_json::Value> = p
        .stops
        .iter()
        .map(|s| {
            serde_json::json!([
                serde_json::to_value(&s.t).unwrap_or(serde_json::Value::Null),
                s.colour.as_array()[0],
                s.colour.as_array()[1],
                s.colour.as_array()[2],
                s.colour.as_array()[3],
            ])
        })
        .collect();
    let mut special_obj = serde_json::Map::new();
    for (k, c) in &p.special {
        let v = serde_json::to_value(k).unwrap_or(serde_json::Value::Null);
        let key = if let Some(s) = v.as_str() {
            s.to_string()
        } else {
            v.to_string()
        };
        special_obj.insert(
            key,
            serde_json::json!([
                c.as_array()[0],
                c.as_array()[1],
                c.as_array()[2],
                c.as_array()[3],
            ]),
        );
    }
    let mut obj = serde_json::Map::new();
    obj.insert("stops".to_string(), serde_json::Value::Array(stops));
    obj.insert(
        "special".to_string(),
        serde_json::Value::Object(special_obj),
    );
    serde_json::Value::Object(obj)
}

fn pos_xy(p: Position) -> serde_json::Value {
    serde_json::Value::Array(vec![
        serde_json::Value::Number(p.x.into()),
        serde_json::Value::Number(p.y.into()),
    ])
}

/// Convert a `Dump` value to its tagged dict representation. Mirrors the
/// `_serialise_dump` function in the Python source.
#[must_use]
pub fn serialise_dump(v: &Dump) -> serde_json::Value {
    match v {
        Dump::BoolGrid { data, palette } => serde_json::json!({
            "$type": "boolgrid",
            "v": data,
            "palette": serialise_palette_t(palette),
        }),
        Dump::U8Grid { data, palette } => serde_json::json!({
            "$type": "u8grid",
            "v": data,
            "palette": serialise_palette_t(palette),
        }),
        Dump::I16Grid { data, palette } => serde_json::json!({
            "$type": "i16grid",
            "v": data,
            "palette": serialise_palette_t(palette),
        }),
        Dump::U16Grid { data, palette } => serde_json::json!({
            "$type": "u16grid",
            "v": data,
            "palette": serialise_palette_t(palette),
        }),
        Dump::F32Grid { data, palette } => serde_json::json!({
            "$type": "f32grid",
            "v": data,
            "palette": serialise_palette_t(palette),
        }),
        Dump::Tiles { data } => serde_json::json!({
            "$type": "tiles",
            "v": data.iter().copied().map(pos_xy).collect::<Vec<_>>(),
        }),
        Dump::Tile { pos } => match pos {
            None => serde_json::json!({"$type": "tile", "x": null, "y": null}),
            Some(p) => serde_json::json!({"$type": "tile", "x": p.x, "y": p.y}),
        },
        Dump::Dot { pos, colour } => {
            let arr = colour.as_array();
            let c = serde_json::json!([arr[0], arr[1], arr[2], arr[3]]);
            match pos {
                None => serde_json::json!({
                    "$type": "dot",
                    "x": null,
                    "y": null,
                    "colour": c,
                }),
                Some(p) => serde_json::json!({
                    "$type": "dot",
                    "x": p.x,
                    "y": p.y,
                    "colour": c,
                }),
            }
        }
        Dump::Path { points, colour } => {
            let arr = colour.as_array();
            serde_json::json!({
                "$type": "path",
                "v": points.iter().copied().map(pos_xy).collect::<Vec<_>>(),
                "colour": [arr[0], arr[1], arr[2], arr[3]],
            })
        }
        Dump::VectorField { angles, magnitudes } => {
            let mut obj = serde_json::Map::new();
            obj.insert("$type".to_string(), serde_json::json!("vectorfield"));
            obj.insert(
                "angles".to_string(),
                serde_json::to_value(angles).unwrap_or(serde_json::Value::Null),
            );
            if let Some(m) = magnitudes {
                obj.insert(
                    "magnitudes".to_string(),
                    serde_json::to_value(m).unwrap_or(serde_json::Value::Null),
                );
            }
            serde_json::Value::Object(obj)
        }
        Dump::Scalar { value } => {
            let v = match value {
                ScalarValue::Int(i) => serde_json::json!(i),
                ScalarValue::Float(f) => serde_json::json!(f),
                ScalarValue::Bool(b) => serde_json::json!(b),
                ScalarValue::Str(s) => serde_json::json!(s),
                ScalarValue::Null => serde_json::Value::Null,
            };
            serde_json::json!({"$type": "scalar", "v": v})
        }
    }
}

/// Auto-wrap a raw value into a tagged dict for use in `debug()` message args.
/// Mirrors the Python `_auto_wrap`. Position values become hoverable cell rings;
/// `Dump` values are used as-is; anything else is a scalar.
#[must_use]
pub fn auto_wrap_position(p: Position) -> serde_json::Value {
    serde_json::json!({"$type": "tile", "x": p.x, "y": p.y})
}

#[must_use]
pub fn auto_wrap_scalar(v: ScalarValue) -> serde_json::Value {
    serialise_dump(&Dump::Scalar { value: v })
}

#[must_use]
pub fn auto_wrap_dump(d: &Dump) -> serde_json::Value {
    serialise_dump(d)
}

// ---------------------------------------------------------------------
// Dumper: per-unit (per-subinterpreter) state and tree builder.
// ---------------------------------------------------------------------

/// Per-unit same-elision dump emitter. Mirrors the Python `Dumper` class.
///
/// Holds a per-name cache of the most recently emitted serialised payload.
/// `dump(name, value)` either appends a fresh node to the current scope's
/// `children`, or emits a `{"$type": "same"}` marker if the payload is
/// byte-identical to the last value emitted under `name`.
///
/// The scope stack is owned externally; `Dumper` only mutates the top frame's
/// `children` list.
pub struct Dumper {
    same_cache: HashMap<String, serde_json::Value>,
}

impl Dumper {
    #[must_use]
    pub fn new() -> Self {
        Self {
            same_cache: HashMap::new(),
        }
    }

    /// Append a vis node to `scope_children`. If the payload equals the last
    /// value emitted under `name`, write a `{"$type": "same"}` marker so the
    /// viewer can reuse the previous turn's value.
    pub fn dump(&mut self, scope_children: &mut Vec<serde_json::Value>, name: &str, value: &Dump) {
        let payload = serialise_dump(value);
        let same = self.same_cache.get(name) == Some(&payload);
        if !same {
            self.same_cache.insert(name.to_string(), payload.clone());
        }
        let value_field = if same {
            serde_json::json!({"$type": "same"})
        } else {
            payload
        };
        let node = serde_json::json!({
            "$type": "vis",
            "name": name,
            "value": value_field,
        });
        scope_children.push(node);
    }
}

impl Default for Dumper {
    fn default() -> Self {
        Self::new()
    }
}

File: bots/beacon/bestrn/util/visualiser.py
"""
Translation of `bots/intgrah/v54.7.9/util/visualiser.py`.

Visualisation primitives. Defines the `Dump` enum (each variant a typed
payload the viewer knows how to render) and the `Dumper` struct that bots
use to emit a per-turn JSON tree of named dump nodes.
"""
from __future__ import annotations

from typing import Final
from dataclasses import dataclass

from typing import TYPE_CHECKING
if TYPE_CHECKING:
    from cambc import Position

class Colour:
    r: int
    g: int
    b: int
    a: int

    def __init__(self, r, g, b, a):
        self.r = r
        self.g = g
        self.b = b
        self.a = a

    def as_array(self):
        return [self.r, self.g, self.b, self.a]
TRANSPARENT: Final[Colour] = Colour(0, 0, 0, 0)

class PaletteStop:
    """One stop in a `Palette`, generic over the scalar type `T` (bool / int / float)."""
    t: T
    colour: Colour

    def __init__(self, t: T, colour: Colour):
        self.t = t
        self.colour = colour

class Palette:
    """Linearly interpolated colour palette over scalar values of type `T`."""
    stops: list[PaletteStop]
    special: list[tuple[T, Colour]]

    def __init__(self, stops, special):
        self.stops = stops
        self.special = special

def green_red():
    return Palette([PaletteStop(t=0, colour=Colour(50, 200, 50, 140)), PaletteStop(t=100, colour=Colour(200, 50, 50, 140))], [])

def blue_red():
    return Palette([PaletteStop(t=0, colour=Colour(50, 50, 200, 140)), PaletteStop(t=100, colour=Colour(200, 50, 50, 140))], [])

def fog():
    return Palette([PaletteStop(t=False, colour=TRANSPARENT), PaletteStop(t=True, colour=Colour(0, 0, 0, 180))], [])

"""A scalar value carried by a `DumpScalar` node."""
@dataclass(frozen=True, slots=True)
class ScalarValueInt:
    _0: int

@dataclass(frozen=True, slots=True)
class ScalarValueFloat:
    _0: float

@dataclass(frozen=True, slots=True)
class ScalarValueBool:
    _0: bool

@dataclass(frozen=True, slots=True)
class ScalarValueStr:
    _0: str

@dataclass(frozen=True, slots=True)
class ScalarValueNull:
    pass

type ScalarValue = ScalarValueInt | ScalarValueFloat | ScalarValueBool | ScalarValueStr | ScalarValueNull

"""
All dumpable payload types. Each variant matches one Python `Dump*`
dataclass; serialisation produces the exact `{"$type": "...", ...}` shape
the Python viewer expects.
"""
@dataclass(frozen=True, slots=True)
class DumpBoolGrid:
    data: list[bool]
    palette: Palette

@dataclass(frozen=True, slots=True)
class DumpU8Grid:
    data: list[int]
    palette: Palette

@dataclass(frozen=True, slots=True)
class DumpI16Grid:
    data: list[int]
    palette: Palette

@dataclass(frozen=True, slots=True)
class DumpU16Grid:
    data: list[int]
    palette: Palette

@dataclass(frozen=True, slots=True)
class DumpF32Grid:
    data: list[float]
    palette: Palette

@dataclass(frozen=True, slots=True)
class DumpTiles:
    """Unordered set of tiles, rendered as cell rings."""
    data: list[Position]

@dataclass(frozen=True, slots=True)
class DumpTile:
    """
    Single tile rendered as a ring around the cell. `pos = None` renders
    as nothing so callers can dump unconditionally.
    """
    pos: Position | None

@dataclass(frozen=True, slots=True)
class DumpDot:
    """Single tile rendered as a filled coloured dot."""
    pos: Position | None
    colour: Colour

@dataclass(frozen=True, slots=True)
class DumpPath:
    """Ordered list of tiles rendered as a polyline."""
    points: list[Position]
    colour: Colour

@dataclass(frozen=True, slots=True)
class DumpVectorField:
    angles: list[float | None]
    magnitudes: list[float] | None

@dataclass(frozen=True, slots=True)
class DumpScalar:
    """Plain value (int / float / bool / str / None) shown verbatim."""
    value: ScalarValue

type Dump = DumpBoolGrid | DumpU8Grid | DumpI16Grid | DumpU16Grid | DumpF32Grid | DumpTiles | DumpTile | DumpDot | DumpPath | DumpVectorField | DumpScalar

def serialise_palette_t(p):
    stops: list[Value] = list(([s.t, s.colour.as_array()[0], s.colour.as_array()[1], s.colour.as_array()[2], s.colour.as_array()[3]] for s in p.stops))
    special_obj = {}
    for k, c in p.special:
        key = str(k)
        special_obj[key] = [c.as_array()[0], c.as_array()[1], c.as_array()[2], c.as_array()[3]]
    obj = {}
    obj[str("stops")] = stops
    obj[str("special")] = special_obj
    return obj

def pos_xy(p):
    return [p.x, p.y]

def serialise_dump(v):
    """
    Convert a `Dump` value to its tagged dict representation. Mirrors the
    `_serialise_dump` function in the Python source.
    """
    match v:
        case DumpBoolGrid(data=data, palette=palette):
            return {"$type": "boolgrid", "v": data, "palette": serialise_palette_t(palette)}
        case DumpU8Grid(data=data, palette=palette):
            return {"$type": "u8grid", "v": data, "palette": serialise_palette_t(palette)}
        case DumpI16Grid(data=data, palette=palette):
            return {"$type": "i16grid", "v": data, "palette": serialise_palette_t(palette)}
        case DumpU16Grid(data=data, palette=palette):
            return {"$type": "u16grid", "v": data, "palette": serialise_palette_t(palette)}
        case DumpF32Grid(data=data, palette=palette):
            return {"$type": "f32grid", "v": data, "palette": serialise_palette_t(palette)}
        case DumpTiles(data=data):
            return {"$type": "tiles", "v": list((pos_xy(__x) for __x in data))}
        case DumpTile(pos=pos):
            return ({"$type": "tile", "x": p.x, "y": p.y} if (p := pos) is not None else {"$type": "tile", "x": None, "y": None})
        case DumpDot(pos=pos, colour=colour):
            arr = colour.as_array()
            c = [arr[0], arr[1], arr[2], arr[3]]
            match pos:
                case None:
                    return {"$type": "dot", "x": None, "y": None, "colour": c}
                case p if p is not None:
                    return {"$type": "dot", "x": p.x, "y": p.y, "colour": c}
        case DumpPath(points=points, colour=colour):
            arr = colour.as_array()
            return {"$type": "path", "v": list((pos_xy(__x) for __x in points)), "colour": [arr[0], arr[1], arr[2], arr[3]]}
        case DumpVectorField(angles=angles, magnitudes=magnitudes):
            obj = {}
            obj[str("$type")] = "vectorfield"
            obj[str("angles")] = angles
            m = magnitudes
            if m is not None:
                obj[str("magnitudes")] = m
            return obj
        case DumpScalar(value=value):
            match value:
                case ScalarValueInt(_0=i):
                    v = i
                case ScalarValueFloat(_0=f):
                    v = f
                case ScalarValueBool(_0=b):
                    v = b
                case ScalarValueStr(_0=s):
                    v = s
                case ScalarValueNull():
                    v = None
            return {"$type": "scalar", "v": v}

def auto_wrap_position(p):
    """
    Auto-wrap a raw value into a tagged dict for use in `debug()` message args.

    Mirrors the Python `_auto_wrap`. Position values become hoverable cell rings;
    `Dump` values are used as-is; anything else is a scalar.
    """
    return {"$type": "tile", "x": p.x, "y": p.y}

def auto_wrap_scalar(v):
    return serialise_dump(DumpScalar(value=v))

def auto_wrap_dump(d):
    return serialise_dump(d)

class Dumper:
    """
    Per-unit same-elision dump emitter. Mirrors the Python `Dumper` class.

    Holds a per-name cache of the most recently emitted serialised payload.
    `dump(name, value)` either appends a fresh node to the current scope's
    `children`, or emits a `{"$type": "same"}` marker if the payload is
    byte-identical to the last value emitted under `name`.

    The scope stack is owned externally; `Dumper` only mutates the top frame's
    `children` list.
    """
    same_cache: dict[str, Value]

    def __init__(self):
        self.same_cache = {}

    def dump(self, scope_children, name, value):
        """
        Append a vis node to `scope_children`. If the payload equals the last
        value emitted under `name`, write a `{"$type": "same"}` marker so the
        viewer can reuse the previous turn's value.
        """
        payload = serialise_dump(value)
        same = self.same_cache.get(name) == payload
        if not same:
            self.same_cache[str(name)] = list(payload)
        value_field = {"$type": "same"} if same else payload
        node = {"$type": "vis", "name": name, "value": value_field}
        scope_children.append(node)

    @staticmethod
    def default():
        return Dumper()
